more_itertools/more.py

```diff
@@ -445,3 +445,81 @@ def windowed(seq, n, fillvalue=None):
     for item in it:
         append(item)
         yield tuple(window)
+
+
+class separate(object):
+    """
+    Wraps an iterable and returns an object that separates the iterable
+    into child iterables based on the *key* function.
+
+    >>> iterable = ['a1', 'b1', 'c1', 'a2', 'b2', 'c2', 'b3']
+    >>> s = separate(iterable, key=lambda s: s[0])  # Select by first character
+    >>> a_iterable = s['a']
+    >>> next(a_iterable)
+    'a1'
+    >>> next(a_iterable)
+    'a2'
+    >>> list(s['b'])
+    ['b1', 'b2', 'b3']
+
+    The original iterable will be advanced and its items will be cached until
+    they are used by the child iterables. This may require significant storage.
+
+    By default, attempting to select a value that does not match any items in
+    the iterable will exhaust the iterable, caching all the values.
+    If a set of *values* is provided, items that don't match any will not be
+    cached, and attempting to select them will generate a KeyError.
+
+    >>> iterable = [0, 1, 2, 3, 4, 5, 6, 7, 8]
+    >>> mod_3 = lambda x: x % 3  # Remainder when divided by 3
+    >>> values = {0, 1}  # Only keep items that equal 0 or 1 (mod 3)
+    >>> s = separate(iterable, key=mod_3, values=values)
+    >>> list(s[0])
+    [0, 3, 6]
+    >>> list(s[1])
+    [1, 4, 7]
+    >>> list(s[2])
+    Traceback (most recent call last):
+    ...
+    KeyError: 2
+
+    """
+
+    def __init__(self, iterable, key, values=tuple()):
+        self._it = iter(iterable)
+        self._key = key
+        self._cache = defaultdict(deque)
+        self._values = set(values)
+
+    def __contains__(self, value):
+        if self._values and (value not in self._values):
+            return False
+
+        try:
+            item = next(self[value])
+        except StopIteration:
+            return False
+        else:
+            self._cache[value].appendleft(item)
+
+        return True
+
+    def _get_values(self, value):
+        while True:
+            if self._cache[value]:
+                yield self._cache[value].popleft()
+            else:
+                while True:
+                    item = next(self._it)
+                    item_value = self._key(item)
+                    if item_value == value:
+                        yield item
+                        break
+                    if self._values and (item_value not in self._values):
+                        continue
```

Comment:
Re: raising an exception for unexpected values... not sure. I think I'm going to fall back to simplicity.